articles/realtime.md

```diff
@@ -0,0 +1,259 @@
+However, it is generally agreed upon that a real-time system is not defined by low latency, but by a deterministic schedule: it must be guaranteed that the system finishes a certain task by a certain time.
+Therefore, it is important that the latency in the system be measurable and a maximum allowable latency for tasks be set.
+
+A more useful metric in evaluating the "hardness" of a real-time system is jitter, which is defined as the variable deviation between a task's deadline and its actual time of completion.
+A more jittery system is less deterministic, less predictable, and less real-time.
+Though in practice it is impossible to completely eliminate jitter from a real-time system, it is a worthy goal to determine a hard upper bound for jitter.
+
+# Implementation of Real-time Computing
+
+In general, an operating system can guarantee that the tasks it handles for the developer, such as thread scheduling, are deterministic, but the OS may not guarantee that the developer's code will run in real-time.
+Therefore, it is up to the developer to know what the determinstic guarantees of an existing system are, and what she must do to write hard real-time code on top of the OS.
+
+In this section, various strategies for developing on top of a real-time OS are explored, since these strategies might be applicable to ROS 2.
+The patterns focus on the use case of C/C++ development on Linux-based real-time OS's (such as RTLinux), but the general concepts are applicable to other platforms.
+
+## Memory management
+
+### Lock memory, prefault stack:
+
+```c
+    if (mlockall(MCL_CURRENT|MCL_FUTURE) == -1) {
+            perror("mlockall failed");
+            exit(-2);
+    }
+    unsigned char dummy[MAX_SAFE_STACK];
+
+    memset(dummy, 0, MAX_SAFE_STACK);
+```
+
+`mlockall` is a Linux system call for locking the process's virtual address space into RAM, preventing the memory that will be accessed by the process from getting paged into swap space.
+Paging and accessing paged memory is a nondeterministic operation that should be avoided in real-time computation.
+
+This code snippet, when run at the beginning of a thread's lifecycle, ensures that no pagefaults occur while the thread is running.
+`mlockall` locks the stack for the thread.
+The `memset` call pre-loads each block of memory of the stack into the cache, so that no pagefaults will occur when the stack is accessed.
+
+### Allocate dynamic memory pool
+
+```c
+  if (mlockall(MCL_CURRENT | MCL_FUTURE))
+    perror("mlockall failed:");
+
+  /* Turn off malloc trimming.*/
+  mallopt(M_TRIM_THRESHOLD, -1);
+
+  /* Turn off mmap usage. */
+  mallopt(M_MMAP_MAX, 0);
+
+  page_size = sysconf(_SC_PAGESIZE);
+  buffer = malloc(SOMESIZE);
+
+  for (i=0; i < SOMESIZE; i+=page_size)
+  {
+      buffer[i] = 0;
+  }
+  free(buffer);
+```
+
+It is commonly believed that dynamic memory allocations are not permitted in the real-time code path.
```

Comment:
Reword 'commonly believed', as it again gives the impression that this is an opinionated statement. If the default allocator is non-deterministic or subject to causing page faults, then it cannot be used in the real-time path. Guidelines for critical system code like [MISRA C](http://www.misra.org.uk/MISRAHome/MISRAC2012/tabid/196/Default.aspx) (paywalled) go as far as recommending not using `malloc`, ever (!!!), because it's too implementation dependent, and very little guarantees can be made on its behavior.